src/v/cluster/topics_frontend.cc

```diff
@@ -841,6 +843,26 @@ ss::future<topic_result> topics_frontend::do_create_topic(
     if (!units) {
         co_return make_error_result(assignable_config.cfg.tp_ns, units.error());
     }
+
+    // Enforce cluster-wide topic count limit only on user created topics.
+    auto max_user_topics_opt = _max_user_topics();
+    if (max_user_topics_opt.has_value()) {
+        // This is intended to approximate the number of topics that are
+        // user-defined and may return a count slightly smaller than the actual
+        // amount.
+        auto user_topic_count = std::max(
+                                  _topics.local().all_topics_count(),
+                                  model::non_user_topics.size())
+                                - model::non_user_topics.size();
+        if (
+          user_topic_count > *max_user_topics_opt
+          && model::is_user_topic(tp_ns)) {
+            co_return make_error_result(
+              assignable_config.cfg.tp_ns,
+              make_error_code(errc::topic_operation_error));
```

Comment:
since the returned error is very generic, should we log a message about why creation failed? also, `kafka::protocol::creatable_topic_response` looks like it contains a free-form string `error_message` that could provide more context to the client?

Reply:
Good idea, will add a message.